components/service/glean/src/main/java/mozilla/components/service/glean/storages/PingStorageEngine.kt

```diff
@@ -0,0 +1,178 @@
+ */
+internal open class PingStorageEngine(context: Context) {
+
+    private val logger: Logger = Logger("glean/PingStorageEngine")
+
+    @VisibleForTesting
+    internal val storageDirectory: File = File(context.applicationInfo.dataDir, PING_STORAGE_DIRECTORY)
+
+    companion object {
+        // Since ping file names are UUIDs, this matches UUIDs for filtering purposes
+        private const val FILE_PATTERN =
+            "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
+        // Base directory for storing serialized pings
+        private const val PING_STORAGE_DIRECTORY = "${Glean.GLEAN_DATA_DIR}/ping_storage"
+    }
+
+    /**
+     * Function to store the ping to file for the [PingUploadWorker] to read and send when it is run
+     *
+     * @param pingType String representing the ping type: baseline, events, metrics (or custom ping)
+     * @param pingPath String representing the upload path used for uploading the ping
+     * @param ping Serialized JSON string representing the ping payload
+     */
+    @ObsoleteCoroutinesApi
+    @Synchronized
+    fun store(pingType: String, uuidFileName: UUID, pingPath: String, ping: String) {
+        Dispatchers.API.launch {
+            val pingStorageDirectory = File(storageDirectory, pingType)
+            assertDirectory(pingStorageDirectory)
+
+            // Build the file path for the ping, using the UUID from the path for the file name
+            val pingFile = File(pingStorageDirectory, uuidFileName.toString())
+
+            var stream: FileOutputStream? = null
+
+            try {
+                stream = FileOutputStream(pingFile, true)
+
+                val writer = BufferedWriter(OutputStreamWriter(stream))
+                writer.write(pingPath)
+                writer.newLine()
+                writer.write(ping)
+                writer.newLine()
+                writer.flush()
+                writer.close()
+            } catch (e: IOException) {
+                logger.warn("IOException while writing event to disk", e)
+            } finally {
+                stream?.close()
+            }
+        }
+    }
+
+    /**
+     * Function used by [PingUploadWorker] to process all serialized ping files for a particular
+     * ping type
+     *
+     * @param pingType String representing the ping type: baseline, events, metrics (or custom ping)
+     * @param callback Callback function to do the actual process action on the ping.  Typically
+     *                 this will be [HttpPingUploader.upload]
+     * @return Boolean representing the success of the upload task. This may be the value bubbled up
+     *         from the callback, or a local value if there was an error reading the files. The
+     *         [PingUploadWorker] uses this to determine whether or not to retry or declare success.
+     */
+    fun process(pingType: String, callback: (String, String) -> Boolean): Boolean {
```

Comment:
Why do we need to process per ping type? What's the advantage over processing all the files in a flat structure?

Reply:
I thought it would be much easier to deal with the ping types (baseline, events, metrics) separately to reduce the IO overhead of a single worker.  I figured the OS would be less likely to kill 3 shorter running worker processes than one longer one.  If this was a single worker process and it exceptioned on a particular ping upload and got cancelled, we may lose or never upload valid pings.  Separating the work across more workers seemed like a way to isolate that and give us a better chance of getting the data to the server.

Reply:
I see. However, we wouldn't be uploading pings until a new ping is enqueued (e.g. when we get to background), right? Moreover, we don't really expect to have tons of pending pings, that would likely be a sign of something going wrong. Anyway, I'd leave this up to you to decide/figure out :)